src/tools/rbd_mirror/image_map/Policy.cc

```diff
@@ -0,0 +1,200 @@
+  if (instance_id == UNMAPPED_INSTANCE_ID) {
+    return false;
+  }
+
+  m_inmap[instance_id].erase(global_image_id);
+  return true;
+}
+
+template <typename I>
+void Policy<I>::update_images_added(const std::string &mirror_uuid,
+                                    ImageIds &&image_ids, Images *images,
+                                    Context *on_finish) {
+  // images we just mapped -- once we start maintaining state
+  // for images, image map update request can selectively
+  // update on-disk map.
+  Images mapped;
+
+  for (auto const &image_id : image_ids) {
+    dout(20) << ": image_id=" << image_id << dendl;
+
+    std::string instance_id = lookup(image_id.global_id);
+    if (instance_id == UNMAPPED_INSTANCE_ID) {
+      instance_id = map(image_id.global_id);
+      mapped.emplace(image_id.id, image_id.global_id, instance_id);
+    }
+
+    m_info[image_id.global_id].peer_uuid.emplace(mirror_uuid);
+    images->emplace(image_id.id, image_id.global_id, instance_id);
+  }
+
+  if (mapped.empty()) {
+    on_finish->complete(0);
+    return;
+  }
+
+  UpdateRequest<I> *req = UpdateRequest<I>::create(
+    m_ioctx, mapped, on_finish);
+  req->send();
+}
+
+template <typename I>
+void Policy<I>::update_images_removed(const std::string &mirror_uuid,
+                                      ImageIds &&image_ids, Images *images,
+                                      Context *on_finish) {
+  // images that need to be purged -- once we start maintaining
+  // state for images, image map remove request can selectively
+  // purge on-disk map.
+  GlobalImageIds purge;
+
+  for (auto const &image_id : image_ids) {
+    dout(20) << ": image_id=" << image_id << dendl;
+
+    std::string instance_id = lookup(image_id.global_id);
+    assert(instance_id != UNMAPPED_INSTANCE_ID);
+
+    images->emplace(image_id.id, image_id.global_id, instance_id);
+
+    auto rm = m_info[image_id.global_id].peer_uuid.erase(mirror_uuid);
+    assert(rm == 1);
+
+    if (m_info[image_id.global_id].peer_uuid.empty()) {
+      unmap(image_id.global_id);
+      m_info.erase(image_id.global_id);
+
+      purge.emplace(image_id.global_id);
+    }
+  }
+
+  if (purge.empty()) {
+    on_finish->complete(0);
+    return;
+  }
+
+  RemoveRequest<I> *req = RemoveRequest<I>::create(
+    m_ioctx, purge, on_finish);
+  req->send();
+}
+
+template <typename I>
+void Policy<I>::init(Context *on_finish) {
+  dout(20) << dendl;
+
+  LoadRequest<I> *req = LoadRequest<I>::create(
+    m_ioctx, &m_inmap, on_finish);
+  req->send();
+}
+
+template <typename I>
+void Policy<I>::shut_down(Context *on_finish) {
+  dout(20) << dendl;
+
+  m_async_op_tracker.wait_for_ops(on_finish);
+}
+
+template <typename I>
+void Policy<I>::update_images(const std::string &mirror_uuid,
+                              ImageIds &&added_image_ids,
+                              ImageIds &&removed_image_ids,
+                              MapSpecRef mapspec, Context *on_finish) {
```

Comment:
I haven't added the shuffle timestamp yet in the update -- will have that along with additional tests in the next update.